Thread reactions

superbryntendo opened the pull request:
<!-- FILL OUT THE BELOW FORM OR YOUR PR WILL BE AUTOMATICALLY CLOSED -->
**Status**
- [x] WIP
- [ ] Ready for review
- [ ] Needs testing

**Deploy after merge (delete what needn't be deployed)**
- api
- hyperion (frontend)
- desktop
- mobile

**Run database migrations (delete if no migration was added)**
YES

**Release notes for users (delete if codebase-only change)**
- Allows users to "like" threads

**Related issues (delete if you don't know of any)**
Closes #

<!-- If there are UI changes please share mobile and desktop screenshots or recordings. -->

superbryntendo committed:
commenting everything out so I can push

brianlovin committed:
Create threadReactions table and indexes

brianlovin committed:
Add loaders

brianlovin committed:
Add model query

brianlovin committed:
Add query resolvers

brianlovin committed:
Add types

brianlovin committed:
Process reputation and flow type mercury

brianlovin committed:
Add analytics tracking

brianlovin committed:
Add bull queues

brianlovin committed:
Add mutations and mutation resolvers

brianlovin committed:
Add athena notificatio type

brianlovin committed:
Properly return threads from mutations

brianlovin committed:
Clean up function formatting

brianlovin committed:
Add graphql mutations

brianlovin committed:
Add sample frontend implementation

brianlovin committed:
API bugfixes

superbryntendo committed:
renders button and count components. needs new glyphs and backend.

superbryntendo committed:
Merge branch 'thread-likes' of github.com:withspectrum/spectrum into thread-likes

brianlovin committed:
Fix flow

superbryntendo committed:
moves likes into threadHeader and updates it to have Thread info rather than only community info.

superbryntendo committed:
static thread header is good. also added likes to threadCards. quite a bit more to do yet.

brianlovin committed:
Split out api loaders and models

brianlovin committed:
Fix imports

brianlovin committed:
Add client side reactions

brianlovin committed:
Merge branch 'thread-likes' of github.com:withspectrum/spectrum into thread-likes

brianlovin committed:
Fix mutation

brianlovin committed:
Fix flow, eslint, peril

brianlovin committed:
Swap icons, add notification to dropdown

brianlovin committed:
Fix reputation processing

superbryntendo committed:
updates threadCard design to handle likes and statuses more simply

mxstbr commented on api/models/threadReaction.js:
```diff
@@ -0,0 +1,134 @@
+import { events } from 'shared/analytics';
+import { trackQueue } from 'shared/bull/queues';
+
+type ReactionType = 'like';
```
nit: this should probably be `ThreadReactionType`?

mxstbr commented on api/models/threadReaction.js:
```diff
@@ -0,0 +1,134 @@
+          },
+        });
+
+        sendThreadReactionNotificationQueue.add({ threadReaction: thisReaction, userId });
```
This is probably not ideal? If I'm reading this correctly, it would mean that if I spam-click the thread reaction button you'll be sent millions of notifications, one for each click. We should only ever send one of 'em?

brianlovin replied:
This is handled in athena. It works by looking up a previous notification of the same type on the same entity in the previous 30mins, then will de-duplicate actors. So the result is that it will mark the notification as "unread" for the thread author, but will not actually insert multiple notifications unless there is a 30 min gap in between.

mxstbr commented on api/types/Thread.js:
```diff
@@ -97,13 +98,24 @@ const Thread = /* GraphQL */ `
 
+  input AddThreadReactionInput {
+    threadId: ID!
+    type: ReactionTypes
```
nit: should be `ThreadReactionTypes`

mxstbr commented on athena/queues/thread-reaction-notification.js:
```diff
@@ -0,0 +1,120 @@
+    debug('get thread');
+
+    // make sure that the person who left the original message still has notification permissions in this thread. We have to check the threadtype to determine if the reaction was left in a story thread or a direct message thread
+    // TODO: In the future we'll want reactions in direct message threads to trigger push notifications, but for now it introduces too much complexity so we just say false
```
this todo doesn't make any sense, you can't react to a dm thread?

mxstbr commented on athena/queues/thread-reaction-notification.js:
```diff
@@ -0,0 +1,120 @@
+    const thread = await getThreadById(updatedNotification.context.id);
+
+    // make sure that the person who left the original message still has notification permissions in this thread. We have to check the threadtype to determine if the reaction was left in a story thread or a direct message thread
+    // TODO: In the future we'll want reactions in direct message threads to trigger push notifications, but for now it introduces too much complexity so we just say false
```
this todo doesn't make any sense, you can't react to a dm thread?

mxstbr commented on api/types/Thread.js:
```diff
@@ -54,6 +54,7 @@ const Thread = /* GraphQL */ `
     attachments: [Attachment]
     watercooler: Boolean
     currentUserLastSeen: Date @cost(complexity: 1)
+    reactions: ReactionData @cost(complexity: 1)
```
I'm going to veto using the same schema as the message reactions because it doesn't let us show which users liked the thing. (you only see it in the notifications, but nobody else can see/check) Let's change the return type here to include that information somehow, (open to ideas) so we can in the future show who liked a thread or not in the UI.

brianlovin replied:
Going to change this to be a separate type `ThreadReactions` which will expose:

```
count: Int
hasReacted: Boolean
reactors: [User]
```

Or something like that - we can actually extend the message reactions to have the same data model, but let's keep message reactions and thread reactions as separately defined types in case they diverge in the future.

mxstbr reviewed (commented):
Overall LGTM, only concern is the `Thread.reactions` schema as noted below. Let's get something better in there that's not like the `Message.reactions` one, then I'm good with the backend side of things here apart from the nits!

brianlovin committed:
Clean up nits

superbryntendo committed:
sets the two states for likes. The titlebar should be display: none until you pass the actionbar probably?

superbryntendo committed:
Merge branch 'thread-likes' of github.com:withspectrum/spectrum into thread-likes

brianlovin committed:
Eslint

brianlovin committed:
Fix flow

brianlovin committed:
Enable the thread header to slide in/out when the thread content is out of view

brianlovin commented:
Took a while, but finally got this working. @mxstbr the one trouble I ran into was debouncing the `handleScroll` method in that last commit - ideally we would debounce or throttle it so we aren't putting a ton of load on the browser just to determine whether or not to show the header (I found a 100ms debounce is probably a good threshold). But I ran into some issues with React removing synthetic events from the event pool. I tried using `event.persist()` from the docs (https://reactjs.org/docs/events.html) but couldn't get it to work - React kept throwing the error:

```
This synthetic event is reused for performance reasons. If you're seeing this, you're accessing the method `target` on a released/nullified synthetic event. This is a no-op function. If you must keep the original synthetic event around, use event.persist(). See https://facebook.github.io/react/docs/events.html#event-pooling for more information.
```

Any chance you could take a look? Without the debounce people's machines are gonna get real real hot.

brianlovin commented:
Also @uberbryn @mxstbr - if you want to pull this and try it out, let me know what you think. In my opinion it's very very distracting to have this header jump in and out every time you change the thread in the inbox view. But it seems to be quite a challenge to work around that since we're doing so much scroll position handling every time a thread is changed.

brianlovin committed:
Fix flow

superbryntendo commented:
Would be cool for it to be more chill when we jump your scroll position as you mentioned, but I also think it's fine as is - it really wouldn't have bothered me other than that you mentioned it. 

Other solution I could think of would be transitioning it from 0 opacity to 1 so the motion isn't jumpy, but I really don't think it's that big of a deal. Would prefer shipping and refining later.

Great work, bud! Thanks for grabbing that!

mxstbr commented:
Will take a look!

mxstbr commented on src/views/thread/container.js:
```diff
@@ -146,6 +156,34 @@ class ThreadContainer extends React.Component<Props, State> {
+    // enable the `bannerIsVisible` state to slide the thread context banner
+    // in from the top of the screen
+    const scrollOffset = e.target.scrollTop;
+    this.setState({ scrollOffset }, () => {
```
We shouldn't be putting the scroll offset (which changes every single frame you scroll) into the state, this is what's causing the issues. This isn't even necessary, just don't put it into the state, compare `scrollOffset` against `threadDetailHeight` and only set the state if it changes. That way you don't even need to debounce which is kinda annoying for this behavior

I'm taking a look at refactoring this!

mxstbr reviewed (changes requested):
Needs an optimistic update on the like click—feels bad rn on slow networks/with the remote server overhead. Will also tackle.

mxstbr committed:
Stop storing scroll offset in state unnecessarily, causing excessive re-renders

@brianlovin see how I made it so we only `setState` once the state
actually changes? This means we barely re-render the React side of
things, which makes this much much much more performant! :ok_hand:

mxstbr commented on api/types/Thread.js:
```diff
@@ -1,5 +1,14 @@
+  type ThreadReactions {
+		count: Int!
+		hasReacted: Boolean
+	}
```
Even though you split 'em up, this is still the same schema as the message reactions? I thought we agreed that we should do this one right from the start (to one that let's us show who liked a thread) so we wouldn't have to migrate to a different one in the future?

brianlovin replied:
Right, see comment below.

mxstbr commented:
> Needs an optimistic update on the like click—feels bad rn on slow networks/with the remote server overhead. Will also tackle.

Nevermind, played around with this a bit and it's a pain in the butt. Let's not do this for now 😅 So the last bit here is coming up with a nicer schema that let's us show who liked a thread, right?

brianlovin commented:
Hm, the schema will be:
```
type ThreadReactions {
  count: Int!
  hasReacted: Boolean
  users: [User]
}
```

Right? We just add a field inside that type. `users` could follow the connection spec, or just be an array of user data (obviously will be problematic when we get to hundreds of reactions on a post). So then it would be:

```
type ReactionUsersConnection {
  pageInfo: PageInfo
  edges: [ReactionUserEdge]
}

type ReactionUserEdge {
  node: User
  cursor: String
}

type ThreadReactions {
  count: Int!
  hasReacted: Boolean
  users: ReactionUsersConnection
}
```

But again, all of this can be added later. Querying would not change except adding the field to the graphql query and extracting it on the frontend with `thread.reactions.users.edges.map(({ node: user }) => ...)`

brianlovin commented:
Let me see if I can handle optimistic update real quick

brianlovin committed:
Transition opacity to smooth jumpiness

brianlovin committed:
Add optimistic udpates for thread likes

brianlovin committed:
Merge branch 'thread-likes' of github.com:withspectrum/spectrum into thread-likes

brianlovin commented:
Optimistic updating is pushed, @mxstbr this is ready for a final review!

mxstbr reviewed (approved):
That schema actually makes a lot of sense. This LGTM, let's ship it?!

brianlovin commented:
LET'S SHIP IT! Gonna put it on alpha, run the migration, test, then do a prod cut.
